LLama/LLamaTemplate.cs

```diff
@@ -101,6 +104,11 @@ public bool AddAssistant
             }
         }
     }
+    
+    /// <summary>
+    /// Get a span to the underlying bytes as specified by Encoding
+    /// </summary>
+    public ReadOnlySpan<byte> TemplateDataBuffer => _result;
```

Comment:
The reason I created `Apply` with the slightly weird use pattern (call it to do work and discover length, call it again to get results) was to ensure the internal `_result` array was never exposed.

This is to avoid a case like this where the buffer might get modified underneath you without you realising:

```csharp
t.Apply()
var template1 = t.TemplateDataBuffer;
t.RemoveAt(1);
t.Apply()
var template2 = t.TemplateDataBuffer;
// template1 was modified by the second call to Apply!
```

If we don't want to accept that risk, then this property needs to be removed for that reason.

On the other hand, if we _do_ want to accept that risk then we could change `Apply()` to return the span immediately, that leads to a much nicer interface!

```csharp
ReadOnlySpan<byte> Apply()
```

It's up to you which you go with. Having seen it in use I think I'd be slightly in favour of option 2 (as long as it is there is a warning in the method docs comments of course).

---

Sidenote: there's also a bug here, although from the above it probably doesn't matter!

`_result` is a big array that contains the template output. However, it's not _all_ valid data. `_resultLength` indicates how much of the array currently contains useful data. So this should be `_result.AsSpan(0, _resultLength);`.

Reply:
I'm in favor of option 2 as well. I'd argue the internal buffer shouldn't be a class member at all. Every call to Apply would be idempotent and just return a new buffer and however callers chose to share or pass around that data is now up to them. I'd also be in favor of just returning the string here as is and not have to force them to also do an encoding translation or at the least an overload that does such

Reply:
Sounds good, made the changes to support option 2.

I guess my gap is the use cases I've seen it's always been applied the last message added (sans the system prompt and user message seeding) likely because of the `InteractiveExecutor`. We'll likely be able to make additional improvements here since the history would be additive so re-templating the entire history on every call seems like overkill. Something that can be explored later.

Reply:
At the moment the templater relies entirely on llama.cpp which just supplies one method that takes a list and applies the template to it, so any kind of incremental updating is difficult with that particular API. I've been wondering about other upstreaming some improvements to llama.cpp to make it incremental, but that might a bit beynd my C++ abilities.